Fix git_config always changed with space in values

kylelaker opened the pull request:
When the current configuration item has in a space in it, `git-config`
will wrap the value in single quotes(`'`). This causes Ansible to
always, sometimes falsely, to think that the value is being changed;
however, it really isn't. Stripping the extra quotes allows Ansible to
make a true comparison.

felixfontein commented:
Thanks for your contribution! Can you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)?

kylelaker commented:
@felixfontein done. Thank you for the quick response!

felixfontein commented on changelogs/fragments/2028-fix-git_config-false-positives-spaces.yml:
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
+  - git_config - When the existing values have spaces, the step will report as changed only if the content changed
```
```suggestion
  - git_config - when the existing values have spaces, the step will report as changed only if the content changed (https://github.com/ansible-collections/community.general/pull/2028).
```

felixfontein commented:
CC @russoz who recently contributed to this module as well.

russoz commented:
Hi @kylelaker , I haven't been able to reproduce the problem you mention. Can you show us exactly what you are doing?

I ran `git config -l` manually, and some lines for which the values have space in it:
```
alias.unstage=reset HEAD --
alias.last=log -1 HEAD
```
They do not come quoted as you mention.

Then I created a simple playbook to list and add `alias.last` with the existing value:
```yaml
  tasks:
  - name: list git config
    git_config:
      scope: global
      list_all: yes

  - name: add existing
    git_config:
      scope: global
      name: alias.last
      value: log -1 HEAD
      state: present
```
Its output had did not mark it as `changed`:
```
TASK [list git config] *************************************************************************************************************
ok: [localhost] => {"changed": false, "config_values": {"alias.br": "branch", "alias.co": "checkout", "alias.last": "log -1 HEAD", "alias.rb": "rebase", "alias.st": "status", "alias.unstage": "reset HEAD --", "branch.sort": "-committerdate", "core.excludesfile": "/home/az/.gitignore", "diff.tool": "meld", "merge.tool": "meld", "tag.sort": "v:refname", "user.email": "russoz@gmail.com", "user.name": "Alexei Znamensky"}, "msg": ""}

TASK [add existing] ****************************************************************************************************************
ok: [localhost] => {"changed": false, "msg": ""}
```

russoz commented:
I have tried with `local` (and of course `repo`) but also failed to replicate the issue.

felixfontein commented:
Maybe it depends on the git version?

russoz commented:
I'm on 
```
$ git --version
git version 2.25.1
```

felixfontein commented:
I've tested with git 2.31.0, also no single (or double) quotes.

kylelaker commented:
So I am unsure whether this is Ansible or git; however, I am on macOS. 

```
$ uname -a
Darwin Kyles-Mac-mini.local 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:06:51 PST 2021; root:xnu-7195.81.3~1/RELEASE_ARM64_T8101 arm64
$ git --version
git version 2.31.0
$ ansible --version
ansible 2.10.6
  config file = None
  configured module search path = ['/Users/kylelaker/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /opt/homebrew/Cellar/ansible/3.0.0/libexec/lib/python3.9/site-packages/ansible
  executable location = /opt/homebrew/bin/ansible
  python version = 3.9.2 (default, Feb 24 2021, 05:06:40) [Clang 12.0.0 (clang-1200.0.32.29)]
```

And for example:

```
git config --global user.name
'Kyle Laker'
```

All values in `~/.gitconfig` were set using https://github.com/kylelaker/dotfiles/blob/main/roles/git/tasks/main.yml and running this always results in surrounding quotes on this setup. It doesn't seem that I can reproduce on Linux. I don't know if I'd be able to test on Windows.

kylelaker commented:
And for completeness:

```
$ rm ~/.gitconfig
$ ansible-playbook -i hosts -c local local.yml

PLAY [Configure dotfiles] *************************************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************************
[WARNING]: Platform darwin on host localhost is using the discovered Python interpreter at /usr/bin/python, but future installation of another
Python interpreter could change the meaning of that path. See
https://docs.ansible.com/ansible/2.10/reference_appendices/interpreter_discovery.html for more information.
ok: [localhost]

TASK [common : Create local clone] ****************************************************************************************************************
ok: [localhost]

TASK [common : Create necessary directories] ******************************************************************************************************
ok: [localhost] => (item=~/.config)
ok: [localhost] => (item=~/.local)
ok: [localhost] => (item=~/.local/bin)
ok: [localhost] => (item=~/.local/share)
ok: [localhost] => (item=~/.cache)

TASK [git : Set gitconfig aliases] ****************************************************************************************************************
changed: [localhost] => (item={'name': 'co', 'value': 'checkout'})
changed: [localhost] => (item={'name': 'current', 'value': 'rev-parse --abbrev-ref HEAD'})
changed: [localhost] => (item={'name': 'count', 'value': '!git shortlog -sn'})
changed: [localhost] => (item={'name': 'syncm', 'value': '!git pull upstream master && git push origin master'})
changed: [localhost] => (item={'name': 'sync', 'value': '!CURRENT=$(git current) && git pull upstream $CURRENT && git push origin $CURRENT'})

TASK [git : Set git color settings] ***************************************************************************************************************
changed: [localhost] => (item={'name': 'diff', 'value': 'auto'})
changed: [localhost] => (item={'name': 'status', 'value': 'auto'})
changed: [localhost] => (item={'name': 'branch', 'value': 'auto'})
changed: [localhost] => (item={'name': 'ui', 'value': 'true'})

TASK [git : Set git core settings] ****************************************************************************************************************
changed: [localhost] => (item={'name': 'excludesfile', 'value': '~/.gitignore'})
changed: [localhost] => (item={'name': 'editor', 'value': 'nvim'})

TASK [git : Set git user settings] ****************************************************************************************************************
changed: [localhost] => (item={'name': 'name', 'value': 'Kyle Laker'})
changed: [localhost] => (item={'name': 'email', 'value': 'kyle@laker.email'})

TASK [git : Set push settings] ********************************************************************************************************************
changed: [localhost] => (item={'name': 'default', 'value': 'simple'})

TASK [git : Set various git settings] *************************************************************************************************************
changed: [localhost] => (item={'name': 'apply.whitespace', 'value': 'nowarn'})
changed: [localhost] => (item={'name': 'mergetool.keepBackup', 'value': 'false'})
changed: [localhost] => (item={'name': 'difftool.prompt', 'value': 'false'})

TASK [git : Install global gitignore] *************************************************************************************************************
ok: [localhost]

TASK [git : Set PGP signing values] ***************************************************************************************************************
changed: [localhost] => (item={'name': 'user.signingKey', 'value': '0x17CAD266B70E3F72'})
changed: [localhost] => (item={'name': 'commit.pgpSign', 'value': 'true'})

TASK [ssh : Create "~/.ssh" directory] ************************************************************************************************************
ok: [localhost]

TASK [ssh : Install ssh config] *******************************************************************************************************************
ok: [localhost]

TASK [ssh : Accept correct key] *******************************************************************************************************************
ok: [localhost]

TASK [vim : Set variables for vim] ****************************************************************************************************************
skipping: [localhost]

TASK [vim : Set variables for nvim] ***************************************************************************************************************
ok: [localhost]

TASK [vim : Create necessary directories for nvim] ************************************************************************************************
ok: [localhost] => (item=~/.config/nvim/autoload)
ok: [localhost] => (item=~/.config/nvim/undodir)

TASK [vim : Install vim-plug] *********************************************************************************************************************
ok: [localhost]

TASK [vim : Install vimrc] ************************************************************************************************************************
ok: [localhost]

TASK [vim : Install vim plugins] ******************************************************************************************************************
changed: [localhost]

TASK [vim : Update vim plugins] *******************************************************************************************************************
changed: [localhost]

TASK [pyenv : Install pyenv] **********************************************************************************************************************
ok: [localhost]

TASK [pyenv : Install pyenv-virtualenv plugin] ****************************************************************************************************
ok: [localhost]

TASK [pyenv : Install pyenv-update plugin] ********************************************************************************************************
ok: [localhost]

TASK [pyenv : Install pyenv-doctor plugin] ********************************************************************************************************
ok: [localhost]

TASK [zsh : Create zsh config directory] **********************************************************************************************************
ok: [localhost]

TASK [zsh : Create zsh history directory] *********************************************************************************************************
ok: [localhost]

TASK [zsh : Install zshenv] ***********************************************************************************************************************
ok: [localhost]

TASK [zsh : Install zsh config files] *************************************************************************************************************
ok: [localhost] => (item=aliases)
ok: [localhost] => (item=asciinema)
ok: [localhost] => (item=colors)
ok: [localhost] => (item=functions)
ok: [localhost] => (item=loadtools)
ok: [localhost] => (item=pre-ohmyzsh)
ok: [localhost] => (item=wslconfig)
ok: [localhost] => (item=wslfixes)
ok: [localhost] => (item=xdg)
ok: [localhost] => (item=zsh)

TASK [zsh : Install zshrc] ************************************************************************************************************************
ok: [localhost]

TASK [zsh : Install oh-my-zsh] ********************************************************************************************************************
ok: [localhost]

TASK [zsh : Install zsh_syntax_highlighting] ******************************************************************************************************
ok: [localhost]

PLAY RECAP ****************************************************************************************************************************************
localhost                  : ok=31   changed=9    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0

$ ansible-playbook -i hosts -c local local.yml

PLAY [Configure dotfiles] *************************************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************************
[WARNING]: Platform darwin on host localhost is using the discovered Python interpreter at /usr/bin/python, but future installation of another
Python interpreter could change the meaning of that path. See
https://docs.ansible.com/ansible/2.10/reference_appendices/interpreter_discovery.html for more information.
ok: [localhost]

TASK [common : Create local clone] ****************************************************************************************************************
ok: [localhost]

TASK [common : Create necessary directories] ******************************************************************************************************
ok: [localhost] => (item=~/.config)
ok: [localhost] => (item=~/.local)
ok: [localhost] => (item=~/.local/bin)
ok: [localhost] => (item=~/.local/share)
ok: [localhost] => (item=~/.cache)

TASK [git : Set gitconfig aliases] ****************************************************************************************************************
ok: [localhost] => (item={'name': 'co', 'value': 'checkout'})
changed: [localhost] => (item={'name': 'current', 'value': 'rev-parse --abbrev-ref HEAD'})
changed: [localhost] => (item={'name': 'count', 'value': '!git shortlog -sn'})
changed: [localhost] => (item={'name': 'syncm', 'value': '!git pull upstream master && git push origin master'})
changed: [localhost] => (item={'name': 'sync', 'value': '!CURRENT=$(git current) && git pull upstream $CURRENT && git push origin $CURRENT'})

TASK [git : Set git color settings] ***************************************************************************************************************
ok: [localhost] => (item={'name': 'diff', 'value': 'auto'})
ok: [localhost] => (item={'name': 'status', 'value': 'auto'})
ok: [localhost] => (item={'name': 'branch', 'value': 'auto'})
ok: [localhost] => (item={'name': 'ui', 'value': 'true'})

TASK [git : Set git core settings] ****************************************************************************************************************
changed: [localhost] => (item={'name': 'excludesfile', 'value': '~/.gitignore'})
ok: [localhost] => (item={'name': 'editor', 'value': 'nvim'})

TASK [git : Set git user settings] ****************************************************************************************************************
changed: [localhost] => (item={'name': 'name', 'value': 'Kyle Laker'})
ok: [localhost] => (item={'name': 'email', 'value': 'kyle@laker.email'})

TASK [git : Set push settings] ********************************************************************************************************************
ok: [localhost] => (item={'name': 'default', 'value': 'simple'})

TASK [git : Set various git settings] *************************************************************************************************************
ok: [localhost] => (item={'name': 'apply.whitespace', 'value': 'nowarn'})
ok: [localhost] => (item={'name': 'mergetool.keepBackup', 'value': 'false'})
ok: [localhost] => (item={'name': 'difftool.prompt', 'value': 'false'})

TASK [git : Install global gitignore] *************************************************************************************************************
ok: [localhost]

TASK [git : Set PGP signing values] ***************************************************************************************************************
ok: [localhost] => (item={'name': 'user.signingKey', 'value': '0x17CAD266B70E3F72'})
ok: [localhost] => (item={'name': 'commit.pgpSign', 'value': 'true'})

TASK [ssh : Create "~/.ssh" directory] ************************************************************************************************************
ok: [localhost]

TASK [ssh : Install ssh config] *******************************************************************************************************************
ok: [localhost]

TASK [ssh : Accept correct key] *******************************************************************************************************************
ok: [localhost]

TASK [vim : Set variables for vim] ****************************************************************************************************************
skipping: [localhost]

TASK [vim : Set variables for nvim] ***************************************************************************************************************
ok: [localhost]

TASK [vim : Create necessary directories for nvim] ************************************************************************************************
ok: [localhost] => (item=~/.config/nvim/autoload)
ok: [localhost] => (item=~/.config/nvim/undodir)

TASK [vim : Install vim-plug] *********************************************************************************************************************
ok: [localhost]

TASK [vim : Install vimrc] ************************************************************************************************************************
ok: [localhost]

TASK [vim : Install vim plugins] ******************************************************************************************************************
changed: [localhost]

TASK [vim : Update vim plugins] *******************************************************************************************************************
changed: [localhost]

TASK [pyenv : Install pyenv] **********************************************************************************************************************
ok: [localhost]

TASK [pyenv : Install pyenv-virtualenv plugin] ****************************************************************************************************
ok: [localhost]

TASK [pyenv : Install pyenv-update plugin] ********************************************************************************************************
ok: [localhost]

TASK [pyenv : Install pyenv-doctor plugin] ********************************************************************************************************
ok: [localhost]

TASK [zsh : Create zsh config directory] **********************************************************************************************************
ok: [localhost]

TASK [zsh : Create zsh history directory] *********************************************************************************************************
ok: [localhost]

TASK [zsh : Install zshenv] ***********************************************************************************************************************
ok: [localhost]

TASK [zsh : Install zsh config files] *************************************************************************************************************
ok: [localhost] => (item=aliases)
ok: [localhost] => (item=asciinema)
ok: [localhost] => (item=colors)
ok: [localhost] => (item=functions)
ok: [localhost] => (item=loadtools)
ok: [localhost] => (item=pre-ohmyzsh)
ok: [localhost] => (item=wslconfig)
ok: [localhost] => (item=wslfixes)
ok: [localhost] => (item=xdg)
ok: [localhost] => (item=zsh)

TASK [zsh : Install zshrc] ************************************************************************************************************************
ok: [localhost]

TASK [zsh : Install oh-my-zsh] ********************************************************************************************************************
ok: [localhost]

TASK [zsh : Install zsh_syntax_highlighting] ******************************************************************************************************
ok: [localhost]

PLAY RECAP ****************************************************************************************************************************************
localhost                  : ok=31   changed=5    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0
```

russoz commented:
My previous example was on a Linux, but here it is on the Mac:
```
$ git config -l | grep alias
alias.co=checkout
alias.br=branch
alias.st=status
alias.rb=rebase
alias.unstage=reset HEAD --
alias.last=log -1 HEAD
$ git --version
git version 2.30.1
```

russoz commented:
OTOH, calling `strip("'")` will do no harm to the unquoted cases. I just find it weird that we are unable to replicate the issue. Maybe you have some configuration in your system that enables that? I wouldn't know what config value that would be.

In any case, LGTM.

felixfontein commented:
The only way I can get git to print single quotes is by adding the single quotes to the actual config value. @kylelaker can you `cat ~/.gitconfig` and look at what it says for that value? For me,
```
[user]
	email = ...
	name = Felix Fontein
```
without single quotes. If I modify it to
```
[user]
	email = ...
	name = 'Felix Fontein'
```
and run `git config --global user.name`, I actually get `'Felix Fontein'` with single quotes. But I only get them when the single quotes are present in `~/.gitconfig` itself.

russoz commented:
I think I know what the problem is: #1953

russoz commented:
Although now that I take a closer look at that one, it seems to have been solved previously. @kylelaker if you erase your configs and create them again using ansible, can you confirmed they get created *with* quotes?

djmattyg007 commented:
I don't think we should be stripping quotes from the values retrieved from `git config`, because those quotes may be legitimate characters stored in the config file.

Is it possible this is a locale issue or something along those lines? Alternatively, is it an issue with specific versions of git?

kylelaker commented:
Oh yeah, running using `main`, instead of just using the builtin modules in Ansible 2.10, totally works. This doesn't seem to be necessary.

felixfontein commented:
@kylelaker @russoz @djmattyg007 thanks for resolving this!

timblaktu commented:
@kylelaker @russoz @djmattyg007 @felixfontein I'm confused what the resolution to this was. It appears that it resolved for people when they upgraded ansible and/or the community.general collection, however I'm experiencing the same issue, using:

(Note: I am using the community.general collection installed by ansible core, via `poetry add ansible`.)

```
[LAPTOP-9M95256K]tim/src/timblaktu-cm/ansible > poetry run ansible --version
ansible [core 2.12.6]
  config file = /home/tim/src/timblaktu-cm/ansible/ansible.cfg
  configured module search path = ['/home/tim/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/tim/.cache/pypoetry/virtualenvs/timblaktu-cm-ansible-X34RCiD6-py3.10/lib/python3.10/site-packages/ansible
  ansible collection location = /home/tim/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/tim/.cache/pypoetry/virtualenvs/timblaktu-cm-ansible-X34RCiD6-py3.10/bin/ansible
  python version = 3.10.0 (default, Oct 12 2021, 11:25:19) [GCC 8.3.0]
  jinja version = 3.1.2
  libyaml = True
```

which includes community.general version 4.8.2, which appears to be long after the underlying fix to the issue was released. 

```
[LAPTOP-9M95256K]tim/src/timblaktu-cm/ansible > poetry run ansible-galaxy collection list

# /home/tim/.cache/pypoetry/virtualenvs/timblaktu-cm-ansible-X34RCiD6-py3.10/lib/python3.10/site-packages/ansible_collections
Collection                    Version
----------------------------- -------
.
.                                                                                                                                                                                                                                                                                                              
community.general             4.8.2
.
.
```

felixfontein commented:
I think it should have been fixed by #1794 if I understand this issue and the ones linked to correctly.

felixfontein commented:
If I run `ansible localhost -m git_config -a 'scope=global name=user.name value="Felix Fontein"'` locally I don't see any change. `git config --global --get user.name` also returns my name without quotes.

What happens on your machine, and what does `~/.gitconfig` actually contain?

timblaktu commented:
My apologies @felixfontein. I had been installing an older version of community.general collection via a requirements file, which was masking the newer version installed via ansible core. I am no longer seeing the issue.
